keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+behavior from `Disabled` to `Auto` with a new feature gate.
+
+#### Interoperability
+
+If both this annotation and `topologyKeys` are set on a Service, `topologyKeys`
+will be given precedence and this annotation will be ignored.
+
+#### Feature Gate
+
+This functionality will be guarded by the `TopologyAwareHints` feature
+gate. This gate should not be enabled if the `ServiceTopology` gate is enabled.
+
+### API
+
+A new `TopologyHints` struct would be added to the `EndpointSlice.Endpoint`
+struct:
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
+}
+```
+
+```go
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+  // zone indicates the zone this endpoint should be consumed by to enable
+  // topology aware routing.
+  // +optional
+  Zone string `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+}
+```
+
+#### Future API Expansion
+This approach would allow for future API expansion that enabled specifying
+multiple zones per endpoint with weights. That level of complexity may never be
+necessary, but it will be possible. For example:
+
+```yaml
+topologyHints:
+  zones:
+  - name: example-1a
+    weight: 50
+  - name: example-2a
+    weight: 50
+```
+
+Additionally we could easily expand this API to include support for region
+hints. Although it is unclear if either expansion will be necessary, the API is
+designed in a way to make expansions straightforward.
+
+### Kube-Proxy
+
+When the `TopologyAwareHints` feature gate is enabled, Kube-Proxy will be
+updated to filter endpoints based on topology hints when the following
+conditions are true:
+
+- Kube-Proxy is able to determine the zone it is running within (likely based
+  on node labels).
+- The `topology.kubernetes.io/topology-aware-routing` annotation is set to
+  `Auto` for the Service.
+- At least one endpoint for the Service has a hint pointing to the zone
+  Kube-Proxy is running within.
+- All endpoints for the Service have zone hints.
+
+When the above conditions are true, kube-proxy will only route traffic to
+endpoints with a hint referring to the zone Kube-Proxy is running within.
+
+### EndpointSlice Controller
+
+When the `TopologyAwareHints` feature gate is enabled and the
+`topology.kubernetes.io/topology-aware-routing` annotation is set to `Auto`
+for a Service, the EndpointSlice controller will add hints to EndpointSlices.
+These hints will indicate where an endpoint should be consumed by proxy
+implementations to enable topology aware routing.
+
+The EndpointSlice controller will determine how many endpoints should be
+available for each zone based on the proportion of CPU cores in each zone. If
+it is not possible to determine the number CPU cores, 1 core per node will be
+assumed for calculations.
```

Comment:
how does the controller gets this information?
is there a cloud provider interface or is present in the node object?

Reply:
I see you mentioned before 
>- Kube-Proxy is able to determine the zone it is running within (likely based
  on node labels). 
is this the same mechanism?

Reply:
This relies on nodes having the standard topology labels (`topology.kubernetes.io/zone` and `topology.kubernetes.io/region`). I can clarify that here. Here's how the controller gets the info right now: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/utils.go#L52-L63.

Reply:
Added a new section to make this clearer.